lightwood/encoders/text/pretrained.py

```diff
@@ -0,0 +1,362 @@
+        """
+        Given a model, train for n_epochs.
+
+        model - torch.nn model;
+        dataset - torch.DataLoader; dataset to train
+        device - torch.device; cuda/cpu
+        log - lightwood.logger.log; print output
+        optim - transformers.optimization.AdamW; optimizer
+        n_epochs - number of epochs to train
+
+        """
+        self._model.train()
+
+        if optim is None:
+            print("Setting all model params to AdamW")
+            optim = AdamW(self._model.parameters(), lr=5e-5)
+
+        for epoch in range(n_epochs):
+            total_loss = 0
+
+            for batch in dataset:
+                optim.zero_grad()
+
+                inpids = batch["input_ids"].to(self.device)
+                attn = batch["attention_mask"].to(self.device)
+                labels = batch["labels"].to(self.device)
+                outputs = self._model(inpids, attention_mask=attn, labels=labels)
+                loss = outputs[0]
+
+                total_loss += loss.item()
+
+                loss.backward()
+                optim.step()
+                if scheduler is not None:
+                    scheduler.step()
+
+            # self._train_callback(epoch, loss.item())
+            print("Epoch=", epoch + 1, "Loss=", total_loss/len(dataset))
+
+    def _train_callback(self, epoch, loss):
+        log.info(f"{self.name} at epoch {epoch+1} and loss {loss}!")
+
+    def encode(self, column_data):
+        """
+        TODO: Maybe batch the text up; may take too long
+        Given column data, encode the dataset
+        Tokenizer should have a length cap!!
+
+        Args:
+        column_data:: [list[str]] list of text data in str form
+
+        Returns:
+        encoded_representation:: [torch.Tensor] N_sentences x Nembed_dim
+        """
+        encoded_representation = []
+
+        # Freeze training mode while encoding
+        self._model.eval()
+
+        with torch.no_grad():
+            # Set the weights; this is GPT-2
+            for text in column_data:
+
+                # Omit NaNs
+                if text == None:
+                    text = ""
+
+                # Tokenize the text with the built-in tokenizer.
+                inp = self._tokenizer.encode(
+                    text, truncation=True, return_tensors="pt"
+                ).to(self.device)
+
+                output = self._model.base_model(inp).last_hidden_state[:, 0]
+
+                # If the model has a pre-classifier layer, use this embedding.
+                if hasattr(self._model, "pre_classifier"):
+                    output = self._model.pre_classifier(output)
+
+                encoded_representation.append(output.detach())
+
+        return torch.stack(encoded_representation).squeeze(1)
+
+    def decode(self, encoded_values_tensor, max_length=100):
+        raise Exception("Decoder not implemented yet.")
+
+    # @staticmethod
+    # def _mean_norm(xinp, dim=1):
```

Comment:
I'm curious, is this `_mean_norm` method used in the benchmarks you've made so far? How has it performed compared to the disaggregated embeddings?

Reply:
I actually swapped to CLS. Clustering of the the embedded space and a quick inspection of the pretrained only with mean-norm + CLS suggest there is little difference. To follow the fine-tuning prescription, I used the CLS token as the input to the pre-classifier layer